lib/caching.js

```diff
@@ -64,6 +66,58 @@ var caching = function(args) {
         });
     }
 
+    function handleBackgroundRefresh(key, work, options) {
+        if (!backgroundQueue.has(key)) {
+            backgroundQueue.add(key);
+            self.checkRefreshThreshold(key, function(err, isExpiring) {
+                if (err) {
+                    backgroundQueue.delete(key);
+                    return;
+                }
+                if (isExpiring) {
+                    work(function(err, data) {
+                        if (err || !self._isCacheableValue(data)) {
+                            backgroundQueue.delete(key);
+                            return;
+                        }
+                        if (options && typeof options.ttl === 'function') {
+                            options.ttl = options.ttl(data);
+                        }
+                        self.store.set(key, data, options, function() {
+                            backgroundQueue.delete(key);
+                        });
+                    });
+                }
+            });
+        }
+    }
+
+    /**
+     * Checks if the current key is expiring. I.e., if a refreshThreshold is set for this cache
+     * and if the cache supports the ttl method, this methods theck if the remaining ttl is
+     * less than the refreshThreshold.
+     * In all other cases this method callbacks will contain "false" (ie. not expiring).
+     *
+     * @function
+     * @name checkRefreshThreshold
+     *
+     * @param {string} key - The cache key to check.
+     * @param {function} cb
+     *
+     */
+    self.checkRefreshThreshold = function(key, cb) {
+        if (self.refreshThreshold && typeof self.store.ttl === 'function') {
+            return self.store.ttl(key, function(ttlErr, ttl) {
```

Comment:
I assume we don't care about the `ttlErr` here since it's not checked?  Also if `ttl` is `undefined` or `null`, it might result in strange behavior when compared to `self.refreshThreshold`.

Reply:
I added a check for `ttlErr` and for the type of `ttl`. My assumption is that this is done "best effort". If something goes wrong the refresh is not performed.
We may implement some sort of circuit breaking mechanism to avoid overloading the store but I think that it would be out of scope for this PR.

Reply:
Agree, looks good now.